### DISCUSSION TITLE
Probabilistic VS Machine Learning approaches for Record Linkage

### DISCUSSION THREAD

**Discussion by clement-docteur** [moj-analytical-services/splink#2690] [Q&A] [answered]
Hi everyone, I was considering the different approaches for Record Linkage and came across a paper which suggest that machine learning methods (like neural networks) would give better results compared to probabilistic methods.
Here's the paper: ["Beyond Probabilistic Record Linkage: Using Neural Networks and Complex Features to Improve Genealogical Record Linkage" from D. Randall Wilson](https://www.researchgate.net/profile/D-Wilson-2/publication/270285051_wilsonijcnn2011beyondprl/data/54a6e0410cf267bdb909fba6/wilsonijcnn2011beyondprl.pdf)
However, I do realize that most serious attempts at record linkage are using a probabilistic approach, Splink included.

So I'm wondering what do you think about using neural networks or other machine learning algorithms for record linkage? Are they worth considering or is there a reason why most people use a probabilistic method?

Thanks!

**Comment by RobinL** [comment, marked as answer]
Great question.  

My overall sense is that ML methods are probably capable of achieving higher accuracy, likely at the expense of being slower.  However, I think the accuracy gains are probably fairly marginal because a well-specified Fellegi Sunter model is already using most of the available information.  I should not this is more a gut feeling rather than something I can substantiate quantitatively.

So in a nutshell, Splink doesn't claim to offer the absolute highest possible accuracy; rather in most cases, it makes a good tradeoff between accuracy and performance (inference speed).


A few more details:

- The main way a model may achieve higher accuracy would be to somehow make better use of the available information.  

- But often the problem with record linkage is that there simply isn't much information available.  For example, if we're comparing two John Smiths and all we have is their name and birth date, then we wouldn't expect the model to be highly accurate:  it has no way of distinguishing between two different John Smiths who happen to share the same birthday, and the records beloning to the same John Smith.

- I have tried to describe the various _types_ of information that a record linkage model can use as the basis of its prediction [here](https://www.robinlinacre.com/fellegi_sunter_accuracy/).  Broadly speaking, I think a well-specified Fellegi Sunter model with accurate parameter estimates does a pretty good job of using all three information sources.  As a result, we shouldn't expect a ML model to be able to perform much better.  

- That said, the Fellegi Sunter model uses very few parameters.  Counting up the m and u probabilities, you may only have 30 parameters in your whole model.  So it's plausible a ML model with hundreds or thousands of parameters may be able to do better.  

- However, you need to be able to train your ML model accurately, and one problem is often that you don't have lots of training data.  There's not point in having thousands of parameters at your disposal if you have no data to train them on!

- There are some specific areas in which an alternative model may work quite a bit better:

    - The FS model includes a conditional indepdence assumption (see footnote 2 [here](https://www.robinlinacre.com/maths_of_fellegi_sunter/#user-content-fn-2).  If you're modelling data that has strong correlations between columns, then a model that can specifically cater for this is likely to outperform Fellegi Sunter.  You can think about this as the FS making _poor_ use of the information available to it, essentially because it will double count information causing poor predictions.

    - In some cases, there's a _fourth_ source of information over and above the three in the [article i linked above](https://www.robinlinacre.com/fellegi_sunter_accuracy/).  This forth source of information is there may be some constraint on the results that the FS model is unaware of.  An example would be that each record in dataset A can match _at most one_ record in dataset B.  The FS model does not take account of these constraints.  One possibility is to use FS to get probabilities and then apply some additional algorithm to your results to 'use' the information in this constraint.

    - In the FS model it can be quite hard to properly 'extract' information from some _types_ of data.  Specifically, if a column contains a long string (such as a product description), then simple fuzzy matching functions don't work well.  You can use semantic embeddings with cosine similarity within a Splink model.  But if all you have is a description, and you don't have other columns (such as price, SKU, manufacturer and so on), then Splink might not be the best choice.


    Hope that helps.  Obviously I'm biased here so would be interested in others' thoughts.

**Reply by clement-docteur** [reply]
Hi Robin,

Thanks for the detailed and nuanced answer.
As you say, in the end it's all about making better use of the available information. Very enlightening. 

Your answer makes me want to try both approaches, starting with probabilistic and then finding out if I can improve with machine learning. Hopefully I'll be able to come back here in a few weeks to share my experience.

**Comment by medwar99** [comment]
Having worked with graph-based ML for a while, I'd say there's definitely theoretical approaches which could perform linkage as we may want it. There's plenty of work in graph clustering, community finding etc. which may be of interest. Robin has already given a spot-on answer and covered a lot of the points, especially in regards to possibilities of larger models and the drawbacks in training data and "real world" situations that may make them trickier to use. 

In addition to Robin's answer, I'd like to add the **iterpretability** of the FS model is a significant strength over more complex models. NN insight methods have come a long way in recent years, but the relatively "simple" approach of the FS model allows for good confidence and communication of why links have been made. This is especially the case for downstream end-users who may not be familiar with intricacies of ML methods and would benefit from a clear insight into the linkage.

There's definitely merit in considering where other methods can come into the linkage if we consider the parts of the overall pipeline. We are usually trying to identify possible edges, score these edges, and then obtain a decent partitioning of the graph to obtain nice clusters. Different methods exists for each of these stages, and each has their own pros and cons.
  - perhaps using text embedding models to produce record descriptors that are scored,
  - perhaps leveraging more information about the edge than just `match_weight`  to get nicer cuts beyond connected components (for example the comparison set that produced the edge, or the `match_key`),
  - perhaps using graph-based NNs to define higher-order feature descriptors that identify complex links,
  - leveraging techniques to produce model definitions,
  - community metrics to help evaluate and score identified clusters,
  - perhaps leaning more into active learning frameworks to iteratively refine the models.

In the end though I'm very happy with how Splink currently approaches the problem, and the open-source nature of the package means we can test out ideas very easily.

**Reply by clement-docteur** [reply]
Hi Mike, 

Thanks a lot for your answer.
Definitely, interpretability is a very important criterion.

It's interesting that you mention text embedding models. That's an option I was strongly considering because my project involves comparing people's names in English, Chinese Mandarin, and Cantonese, so it might be an idea worth trying in that case. I'm currently testing the different models.
All your other examples are also very relevant. I'll look up those that I hadn't heard of, it's always good to get new ideas.

In the end it's going to be a trade-off between simplicity/interpretability and accuracy. Only if the accuracy gains are significant enough does it make sense to consider using ML methods.

**Reply by Jo-Lam** [reply]
Hi, (and thanks @RobinL for looping me in!)
I happen to have thought about the English v Cantonese v Mandarin scenarios a little bit (see my work: https://ijpds.org/article/view/2935/6186).  Depends on in what form your names are being collected, there might be opportunity to incorporate tonal information & its distribution to help identify true matches. Also be mindful that the distribution of surnames is vastly different for English v Chinese.

Happy to speak more about it (as a Hong Konger who speaks all three ;)

**Reply by clement-docteur** [reply]
Hi Jo, thanks for your answer and for the link to your excellent paper. It confirms what I had in mind about the challenges of Chinese names linkage.
What I get from it is that I should be using Jyutping or Pinyin with tones when available. I wasn't aware that HK romanisation was less standardized and that's definitely something I'll keep in mind.

The problem lies when there are parts of the dataset in pinyin, in traditional characters, and in HK romanisation for example.
My wishful thinking is that there would exist a library that could identify that 张, 張, Zhang, Chang, Cheung, and Cheong can refer to the same thing, but I don't think such a library exists. Hence why I thought that text embedding and evaluating cosine similarity could tell me that 张, 張, Zhang, Chang, Cheung, and Cheong are close to each other? I have yet to try it out to see if it works. Methods like Levenshtein distance seem mostly pointless in that case.
Maybe you have other methods in mind?

**Reply by Jo-Lam** [reply]
Thanks for your kind comments!

Would be keen to see what you find! My very limited experience with name embeddings is that names are not quite like other words. 

I guess it then depends on the task at hand: whether you are trying to separate out Chinese names v non-Chinese names, or differentiate between (potentially) Chinese names.

For the former, you could potentially differentiate with simple q-grams/tri-grams for different distributions of letter combinations. In a slightly different context, I am experimenting with explicitly defined name characteristics and explore ethnic-based differences (ethnicity as proxy for language, it's poor but the best proxy I have). Followed by PCA, then use each PC/altogether (euclidean/cosine) for linkage.

Very prelim findings from my US-based name record is that Asian names (broad group) tend to have longer, and more vowel-rich names, and more use of space " " and "-" in how they're recorded.

For the latter, some of these differences (etc. Zhang v Cheung) are meaningful for linkage too! Classifying them as more similar than (Zhang vs Zheng) might mess with the comparisons. You could potentially start with separating Pinyin vs Non-Pinyin romanisation - as Pinyin is rule-based. 

The ever challenge is it is not possible to reverse the romanisation with 100% confidence. However, depends on your task, it could be worth it to do it for surnames - with informed distribution about surnames in mind. 
The limit is, as explained a bit in the paper, that representing a jyutping name in pinyin might mean you'd miss the similarities between that name with other jyutping names.
